File: 0015-3sum/0015-3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(), nums.end());

        vector<vector<int>> result;

        int start = 0 ;

        while (start < nums.size() - 2) {
            int l = start + 1;
            int r = nums.size() - 1;

            while (l < r) {
                int curr_sum = nums[start] + nums[l] + nums[r];

                if (curr_sum == 0){
                    result.push_back({nums[start], nums[l], nums[r]});
                    while (l < nums.size() - 1) {
                        if (nums[l + 1] == nums[l]) {
                            l += 1;
                        }
                        else {
                            break;
                        }
                    }
                    l++;

                    while (r > start + 1) {
                        if (nums[r - 1] == nums[r]) {
                            r -= 1;
                        }
                        else {
                            break;
                        }
                    }
                    r--;
                }
                else if (curr_sum < 0) {
                    l += 1;
                }
                else {
                    r -= 1;
                }
            }

            while (start < nums.size() - 2) {
                if (nums[start] == nums[start + 1]) {
                    start += 1;
                }
                else {
                    break;
                }
            }
            start++;
        }

        return result;
    }
};

File: 0015-3sum/0015-3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        result = []

        for i in range(len(nums) - 2):
            l, r = i + 1, len(nums) - 1

            if i > 0 and nums[i - 1] == nums[i]:
                continue

            while l < r:

                curr_sum = nums[i] + nums[l] + nums[r]

                if curr_sum == 0:
                    result.append([nums[i], nums[l], nums[r]])

                    while l + 1 < r and nums[l + 1] == nums[l]:
                        l += 1
                    while r - 1 > l and nums[r - 1] == nums[r]:
                        r -= 1
                    l += 1
                    r -= 1
                elif curr_sum < 0:
                    while l + 1 < r and nums[l + 1] == nums[l]:
                        l += 1
                    l += 1
                else:
                    while r - 1 > l and nums[r - 1] == nums[r]:
                        r -= 1
                    r -= 1

        return result